File: packages/contracts/src/domain/scrobble.ts
// Scrobbling contracts shared between the desktop main process (the scrobbler
// service that submits to Last.fm / ListenBrainz) and the renderer Settings UI.
//
// The renderer NEVER sees the raw Last.fm session key or ListenBrainz token —
// those stay main-only. It reads back only this connection status and writes
// credentials in through dedicated IPC.

/**
 * The scrobbling connection status the Settings UI renders. Carries booleans +
 * the display username only — never the underlying secrets.
 */
export interface ScrobbleStatus {
  /** Master opt-in switch. When false, nothing is submitted. */
  enabled: boolean;
  /** True when a Last.fm session key is stored. */
  lastfmConnected: boolean;
  /** Last.fm display name, when connected. */
  lastfmUsername: string | null;
  /** True when a ListenBrainz user token is stored. */
  listenBrainzConnected: boolean;
  /** Number of plays parked in the in-memory retry queue (failed submissions). */
  pendingCount: number;
}

/** Result of connecting Last.fm via the desktop-auth token exchange. */
export type LastfmConnectResult =
  | { ok: true; username: string | null }
  /** `error` is a short reason key for the UI toast. */
  | { ok: false; error: string };

/** Result of connecting ListenBrainz by validating a user token. */
export type ListenBrainzConnectResult =
  | { ok: true; username: string | null }
  | { ok: false; error: string };

File: crates/shiranami-core/src/models/scrobble.rs
//! Scrobbling contracts, ported from
//! `packages/contracts/src/domain/scrobble.ts`.
//!
//! The renderer never sees the raw Last.fm session key or ListenBrainz token —
//! those stay main-only, in [`crate::store`]. It reads back only the connection
//! status here and writes credentials in through dedicated commands.
//!
//! `LastfmConnectResult` / `ListenBrainzConnectResult` are deliberately **not**
//! ported here. Both are `{ ok: true; … } | { ok: false; error }` unions
//! discriminated on a boolean literal, which neither `serde`'s tagged-enum
//! representations nor `specta` 2.0.0-rc.25 (whose literal datatypes are written
//! but unshipped) can express. They are single-command response shapes with no
//! second consumer, so they belong to `shiranami-integrations` in Phase 12,
//! which can pick a representation alongside the command that returns them.

use serde::{Deserialize, Serialize};
use specta::Type;

/// The scrobbling connection status the Settings UI renders.
///
/// Carries booleans and the display username only — never the secrets behind
/// them.
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct ScrobbleStatus {
    /// Master opt-in switch. When false, nothing is submitted.
    pub enabled: bool,
    /// True when a Last.fm session key is stored.
    pub lastfm_connected: bool,
    /// Last.fm display name, when connected.
    pub lastfm_username: Option<String>,
    /// True when a ListenBrainz user token is stored.
    pub listen_brainz_connected: bool,
    /// Plays parked in the retry queue after a failed submission.
    pub pending_count: u32,
}
